Fix #1125: Update PopoverController from rewards-ios.

iccub opened the pull request:
This change also required small adjustments to calculating contentSize
for Shields and Menu view controllers.

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

<!-- Any useful notes for reviewer explaining how best to test and verify. -->
Test if popup views: shields panel and menu panel are showing correctly in both portrait and landscape.

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

iccub committed:
Fix #1125: Update PopoverController from rewards-ios.

This change also required small adjustments to calculating contentSize
for Shields and Menu view controllers.

kylehickinson commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -157,12 +163,12 @@ class ShieldsViewController: UIViewController, PopoverContentComponent {
-        shieldsView.stackView.layoutIfNeeded()
-        preferredContentSize = CGSize(width: PopoverController.preferredPopoverWidth, height: shieldsView.stackView.bounds.height)
-    }
+//    override func viewDidLayoutSubviews() {
```
This can just be deleted instead of commented out, no?

kylehickinson reviewed (changes requested):
Just the one thing, everything else looks good :)

kylehickinson commented:
Looks good, just waiting on CI

jhreis commented on Client/Frontend/Popover/PopoverContainerView.swift:
```diff
@@ -169,7 +170,7 @@ extension PopoverController {
             
             triangleLayer.path = path.cgPath
-            triangleLayer.mask = shadowMask
+            //            triangleLayer.mask = shadowMask
```
Remove entirely.

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -150,6 +154,31 @@ class PopoverController: UIViewController {
+        case .preferredContentSize, .fixedSize(_):
+            containerView.layoutIfNeeded()
+            if contentController.extendEdgeIntoArrow {
+                contentController.view.frame = containerView.contentView.bounds
```
I think early `break` here is better than nesting `else`:

```
if contentController.extendEdgeIntoArrow {
                contentController.view.frame = containerView.contentView.bounds
                break
}

var rect = containerView.contentView.bounds
switch containerView.arrowDirection {
```

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -159,12 +188,33 @@ class PopoverController: UIViewController {
     private let backgroundOverlayView = UIView()
     
     override func preferredContentSizeDidChange(forChildContentContainer container: UIContentContainer) {
+        if (self.isBeingPresented) {
```
Remove parens: `()`

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -469,8 +536,10 @@ extension PopoverController: UIGestureRecognizerDelegate {
         if let scrollView = otherGestureRecognizer.view as? UIScrollView {
+            return false
+            
             let topInset = scrollView.adjustedContentInset.top
```
This is now unreachable, right? If so, we should remove

iccub committed:
Stop dismissal gesture when reaching edge of the scroll view.

That part of code might come back in the future if we decice to restore
that functionality.
iOS 13 does something similar with new gestural dismissal of
presented modal sheets

iccub committed:
Enable pan to dismiss gesture for all views.

kylehickinson commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -459,41 +527,27 @@ extension PopoverController: UIGestureRecognizerDelegate {
         // Don't allow for pan gesture while table view is in editing mode.
         // This bug only occurs for bookmarks at nested levels, see issue #687.
         if let tableView = otherGestureRecognizer.view as? UITableView, tableView.isEditing {
             gestureRecognizer.cancel()
```
Do you think this is still required? I think we can probably just make the entire function return false

iccub replied:
right, it actually seems to return false by default so can remove whole function

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -150,6 +154,32 @@ class PopoverController: UIViewController {
+            switch containerView.arrowDirection {
+            case .up:
+                rect.origin.y = PopoverUX.arrowSize.height
+                rect.size.height -= PopoverUX.arrowSize.height
```
This happens in both .up and .down? If so, it should be moved outside of the switch.

jhreis commented on Client/Frontend/Popover/PopoverContainerView.swift:
```diff
@@ -121,30 +102,50 @@ extension PopoverController {
-                backgroundViewBottomConstraint?.constant = 0.0
+                popoverMaskView.bodyView.frame = contentView.bounds.with {
+                    $0.origin.y = PopoverUX.arrowSize.height
+                    $0.size.height -= PopoverUX.arrowSize.height
```
Since this happens regardless, we should just put a check inside the `with`:

```
.with {
if case == .up { adjustOrigin }
adjustHeight
```

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -214,6 +265,15 @@ class PopoverController: UIViewController {
+        if contentController.extendEdgeIntoArrow {
+            switch containerView.arrowDirection {
+            case .up:
+                contentController.additionalSafeAreaInsets = UIEdgeInsets(top: PopoverUX.arrowSize.height, left: 0, bottom: 0, right: 0)
```
Again, these are the same. Am I missing something here?

iccub replied:
this one can't be shortened easily, because two insets are changed depending on state, top and bottom

jhreis replied:
Oh, I see, they are adjusting different things 👍

kylehickinson commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -214,6 +262,12 @@ class PopoverController: UIViewController {
+        if contentController.extendEdgeIntoArrow {
+            let topInset = containerView.arrowDirection == .up ? PopoverUX.arrowSize.height : 0
+            
+            contentController.additionalSafeAreaInsets = UIEdgeInsets(top: topInset, left: 0, bottom: PopoverUX.arrowSize.height, right: 0)
```
Previous logic is changed now, `additonalSafeAreaInsets` are only set for top or bottom depending on where the area is, unlike origin/height where height had to be changed regardless

kylehickinson commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -457,43 +522,18 @@ extension PopoverController: UIGestureRecognizerDelegate {
         }
-        return false
+    }
+    func navigationController(_ navigationController: UINavigationController, didShow viewController: UIViewController, animated: Bool) {
```
Just realized this is empty, is this a required delegate function?

iccub replied:
removed it
